python/arcticdb/diagnostics_engine/fragmentation_diagnosis.py

```diff
@@ -0,0 +1,197 @@
+# import necessary libraries
+from IPython.display import display, Markdown
+import matplotlib.pyplot as plt
+from .library_utils import check_and_adapt_library, check_symbol_exists
+
+
+def calculate_and_add_segment_occupancy_info(df):
+    """
+    Calculates the number of data points, rows, and columns for each segment in the DataFrame.
+    """
+    df["data_points"] = (df["end_row"] - df["start_row"]) * (df["end_col"] - df["start_col"])
+    df["rows"] = df["end_row"] - df["start_row"]
+    df["columns"] = df["end_col"] - df["start_col"]
+    return df
+
+
+def calculate_sum_segments_occupancy_info(df):
+    """
+    Calculates the total sum of data points, rows, and columns in the DataFrame.
+    """
+    total_data_points = df["data_points"].sum()
+    total_rows = df["rows"].sum()
+    total_columns = df["columns"].sum()
+    return total_data_points, total_rows, total_columns
+
+
+def calculate_segments_occupancy_maximums(df, total_segment_size, segment_row_size, column_group_size):
+    """
+    Calculates the maximum number of data points, rows, and columns that can be stored in the segments.
+    """
+    max_data_points = total_segment_size * df.shape[0]
+    max_rows = segment_row_size * df.shape[0]
+    max_columns = column_group_size * df.shape[0]
+    return max_data_points, max_rows, max_columns
+
+
+def calculate_fragmentation(
+    total_data_points, total_rows, total_columns, max_data_points, max_rows, max_columns, dynamic_schema
+):
+    """
+    Calculates the overall fragmentation level as well as row-wise and column-wise fragmentation for the data.
+    """
+    fragmentation = float(total_data_points) / max_data_points
```

Comment:
This is still wrong with dynamic schema

Reply:
As discussed, we will not calculate/show the total fragmentation for dynamic Schema. It is more sensible/easy to understand to show only the per row fragmentation. Done the changes.